jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventNodeInstance.java

```diff
@@ -141,7 +151,41 @@ public void removeTimerListeners() {
     }
 
     public EventNode getEventNode() {
-        return (EventNode) getNode();
+        EventNode node;
+        try {
+            node = (EventNode) getNode();
+        } catch (ClassCastException e) {
+            long id = getProcessInstance().getId();
+            long nodeInstanceId = getId();
+            logger.debug("node definition changed for process instance "+id+". EventNode not found on node id: "+getNodeId()+" searching with node instance id: "+nodeInstanceId);
+            node = (EventNode)getNodeByNodeInstanceId(id, nodeInstanceId);
+            if (node == null) {
+                throw e;
+            }
+        }
+        return node;
+    }
+
+    /**
+     * Search node by nodeinstanceid and uniqueId, cause node definition id might change
+     * @param id
+     * @param nodeInstanceId
+     * @return
+     */
+    private Node getNodeByNodeInstanceId(long id, long nodeInstanceId) {
+        Node node = null;
+        RuntimeDataService service = (RuntimeDataService) ServiceRegistry.get().service(ServiceRegistry.RUNTIME_DATA_SERVICE);
+        Collection<NodeInstanceDesc> nodes = service.getProcessInstanceNodeInstanceHistory(id, nodeInstanceId, new QueryContext(0, 1));
```

Comment:
incorrect either. the engine should not access the runtime data service. this goes against the design.